wen-restart/src/last_voted_fork_slots_aggregate.rs

```diff
@@ -0,0 +1,245 @@
+            solana::wen_restart_proto::{
+                LastVotedForkSlotsAggregateRecord, LastVotedForkSlotsRecord,
+            },
+        },
+        solana_gossip::restart_crds_values::RestartLastVotedForkSlots,
+        solana_runtime::{
+            bank::Bank,
+            genesis_utils::{
+                create_genesis_config_with_vote_accounts, GenesisConfigInfo, ValidatorVoteKeypairs,
+            },
+        },
+        solana_sdk::{hash::Hash, signature::Signer, timing::timestamp},
+        std::collections::HashMap,
+    };
+
+    #[test]
+    fn test_aggregate() {
+        solana_logger::setup();
+        let validator_voting_keypairs: Vec<_> =
+            (0..10).map(|_| ValidatorVoteKeypairs::new_rand()).collect();
+        let GenesisConfigInfo { genesis_config, .. } = create_genesis_config_with_vote_accounts(
+            10_000,
+            &validator_voting_keypairs,
+            vec![100; validator_voting_keypairs.len()],
+        );
+        let (_, bank_forks) = Bank::new_with_bank_forks_for_tests(&genesis_config);
+        let root_bank = bank_forks.read().unwrap().root_bank();
+        let root_slot = root_bank.slot();
+        let shred_version = 52;
+        let mut slots_vec = Vec::new();
+        let last_voted_fork = vec![root_slot + 1, root_slot + 2, root_slot + 3];
+        let mut slots_aggregate = LastVotedForkSlotsAggregate::new(
+            root_slot,
+            0.42,
+            root_bank.epoch_stakes(root_bank.epoch()).unwrap(),
+            &last_voted_fork,
+            &validator_voting_keypairs[9].node_keypair.pubkey(),
+        );
+        let mut expected_aggregate_record = HashMap::new();
+        for validator_voting_keypair in validator_voting_keypairs.iter().take(3) {
+            let pubkey = validator_voting_keypair.node_keypair.pubkey();
+            slots_vec.push(
+                RestartLastVotedForkSlots::new(
+                    pubkey,
+                    timestamp(),
+                    &last_voted_fork,
+                    Hash::default(),
+                    shred_version,
+                )
+                .unwrap(),
+            );
+            expected_aggregate_record.insert(
+                pubkey.to_string(),
+                LastVotedForkSlotsRecord {
+                    last_vote_fork_slots: last_voted_fork.clone(),
+                    last_vote_bankhash: Hash::default().to_string(),
+                    shred_version: shred_version as u32,
+                },
+            );
+        }
+        let mut aggregate_record = LastVotedForkSlotsAggregateRecord {
+            received: HashMap::new(),
+        };
+        let result = slots_aggregate.aggregate(slots_vec, &mut aggregate_record);
+        assert_eq!(result.active_percent, 40.0);
+        assert!(result.slots_to_repair.is_empty());
+        assert_eq!(aggregate_record.received, expected_aggregate_record);
+
+        let from = validator_voting_keypairs[4].node_keypair.pubkey();
```

Comment:
instead of `4` maybe `initial_num_active_validators + 1`

Reply:
Done.